unicorn/src/unicorn/zookeeper.cpp

```diff
@@ -42,53 +34,570 @@
+        }
+        throw_watch_event(reply);
+    }
+};
+
+class zookeeper_t::increment_t: public safe<versioned_value_t, get_reply_t, create_reply_t, put_reply_t> {
+    zookeeper_t& parent;
+    path_t path;
+    value_t value;
+public:
+    increment_t(callback::increment wrapped, zookeeper_t& parent, path_t path, value_t value) :
+        safe(std::move(wrapped)),
+        parent(parent),
+        path(std::move(path)),
+        value(std::move(value))
+    {}
+
+    auto run() -> void {
+        if (!value.is_double() && !value.is_int() && !value.is_uint()) {
+            throw error_t(error::unicorn_errors::invalid_type, "invalid value type for increment");
+        }
+        parent.zk.get(path, shared_from_this());
+    }
+
+private:
+    auto on_reply(get_reply_t reply) -> void override {
+        if(reply.rc == ZNONODE) {
+            return parent.zk.create(path, serialize(value), false, false, shared_from_this());
+        } else if(reply.rc) {
+            throw error_t(map_zoo_error(reply.rc), "failed to get node value - {}", zerror(reply.rc));
+        }
+        value_t parsed = unserialize(reply.data);
+        if (reply.stat.numChildren != 0) {
+            throw error_t(error::child_not_allowed, "can not increment node with children");
+        } else if (!parsed.is_double() && !parsed.is_int() && !parsed.is_uint()) {
+            throw error_t(error::invalid_type, "can not increment non-numeric value");
+        } else if (parsed.is_double() || value.is_double()) {
+            value = parsed.to<double>() + value.to<double>();
+        } else {
+            value = parsed.to<int64_t>() + value.to<int64_t>();
+        }
+        parent.zk.put(path, serialize(value), reply.stat.version, shared_from_this());
+    }
+
+    auto on_reply(put_reply_t reply) -> void override {
+        if(reply.rc) {
+            throw error_t(map_zoo_error(reply.rc), "failed to put new node value - {}", zerror(reply.rc));
+        }
+        satisfy(versioned_value_t(value, reply.stat.version));
+    }
+
+    auto on_reply(create_reply_t reply) -> void override {
+        if(reply.rc) {
+            throw error_t(map_zoo_error(reply.rc), "could not create value - {}", zerror(reply.rc));
+        } else {
+            satisfy(versioned_value_t(value, version_t()));
+        }
+    }
+};
+
+class zookeeper_t::lock_t : public safe<bool, create_reply_t, children_reply_t, exists_reply_t, del_reply_t, watch_reply_t> {
+public:
+    struct lock_scope_t: public scope_t {
+        std::shared_ptr<lock_t> parent;
+
+        lock_scope_t() {}
+
+        auto close() -> void override {
+            closed.apply([&](bool& closed){
+                if(!closed) {
+                    on_abort();
+                    closed = true;
+                }
+            });
+        }
+
+        auto on_abort() -> void override {
+            BOOST_ASSERT(parent);
```

Comment:
It is not any warranty by type protocol that `parent` would be ever set, so may be use `if` check?